src/usage.rs

```diff
@@ -1,56 +1,70 @@
 use crate::ToParams;
 
 #[derive(Debug, Clone)]
-pub struct Usage {
-    metrics: Vec<(String, String)>,
+pub struct MetricUsage<'m>(&'m str, &'m str);
+
+impl<'m, M: AsRef<str> + 'm, V: AsRef<str> + 'm> From<&'m (M, V)> for MetricUsage<'m> {
+    fn from((m, v): &'m (M, V)) -> Self {
+        Self(m.as_ref(), v.as_ref())
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Usage<'m>(Vec<MetricUsage<'m>>);
+
+impl<'m, M: AsRef<str> + 'm, V: AsRef<str> + 'm> From<&'m [(M, V)]> for Usage<'m> {
+    fn from(mvs: &'m [(M, V)]) -> Self {
+        Self(mvs.iter().map(|mv| MetricUsage::from(mv)).collect())
+    }
 }
 
-// TODO fix this to stop the wild cloning of strings
-// IMO we should treat Usage as a container itself and share it in multiple calls
-// ie.
-// let u = Usage::new();
-// u.push(("metric1", 10));
-// u.push(("metric2", 20));
-// We should also be able to cache the generated parameters, and turn into the inner container,
-// clone, iterate, etc.
-// An internal Vec should do, as we are unlikely to need direct access to specific entries once we
-// have a usage (we can do it anyway, but searching sequentially). If we needed something more
-// powerful we can always write conversions from Vec/HashMap to Usage.
-impl Usage {
-    /// Creates a `Usage`.
+// Add just enough to be able to access the inner vector for any modifications
+impl<'m> Usage<'m> {
+    /// Creates a `Usage` using the `From` implementation.
     ///
     /// # Examples
     ///
     /// ```
     /// use threescalers::usage::*;
     ///
     /// let mut metrics = Vec::new();
-    /// metrics.push(("metric1", 10));
-    /// metrics.push(("metric2", 20));
-    /// let usage = Usage::new(&metrics);
+    /// metrics.push(("metric1", "10"));
+    /// metrics.push(("metric2", "20"));
```

Comment:
Hey @unleashed I've looked into this, but adding support for `ToString` or any integer type would imply the inner storage would change to be a `Cow` like other modules, since in some cases we would need to allocate the storage for the string. Is that what you are suggesting?

The concern of correctness sounds valid but depends on the answer to the question above.

Reply:
You are right, yes, we'd need something like `Cow`. I think eventually we'll probably end up with that, but I'd be ok with this as is for now, and then add it later on if we feel it is really needed.